src/typecheck/mod.rs

```diff
@@ -248,41 +242,271 @@ where
+
+            if attrs.rec {
+                envs.insert(x.clone(), ty_let.clone());
+            }
+
+            linearizer.retype_ident(lin, x, ty_let.clone());
+            walk(state, envs.clone(), lin, linearizer.scope(), re)?;
+
+            if !attrs.rec {
+                envs.insert(x.clone(), ty_let);
+            }
+
+            walk(state, envs, lin, linearizer, rt)
+        }
+        Term::LetPattern(x, pat, re, rt) => {
+            let ty_let = binding_type(re.as_ref(), &envs, state.table, state.wildcard_vars, false, state.resolver);
+            walk(state, envs.clone(), lin, linearizer.scope(), re)?;
+
+            if let Some(x) = x {
+                linearizer.retype_ident(lin, x, ty_let.clone());
+                envs.insert(x.clone(), ty_let);
+            }
+
+            inject_pat_vars(pat, &mut envs);
+
+            walk(state, envs, lin, linearizer, rt)
+        }
+        Term::App(e, t) => {
+            walk(state, envs.clone(), lin, linearizer.scope(), e)?;
+            walk(state, envs, lin, linearizer, t)
+        }
+        Term::Switch(exp, cases, default) => {
+            cases.values().chain(default.iter()).try_for_each(|case| {
+                walk(state, envs.clone(), lin, linearizer.scope(), case)
+            })?;
+
+            walk(state, envs, lin, linearizer, exp)
+        }
+        Term::RecRecord(stat_map, dynamic, ..) => {
+            for id in stat_map.keys() {
+                let binding_type = binding_type(
+                    stat_map.get(id).unwrap().as_ref(),
+                    &envs,
+                    state.table,
+                    state.wildcard_vars,
+                    false,
+                    state.resolver,
+                );
+                envs.insert(id.clone(), binding_type.clone());
+                linearizer.retype_ident(lin, id, binding_type);
+            }
+
+            stat_map
+                .iter()
+                .map(|(_, t)| t)
+                .chain(dynamic.iter().map(|(_, t)| t))
+                .try_for_each(|t| -> Result<(), TypecheckError> {
+                    walk(state, envs.clone(), lin, linearizer.scope(), t)
+                })
+        }
+        Term::Record(stat_map, _) => {
+            for id in stat_map.keys() {
+                let binding_type = binding_type(
+                    stat_map.get(id).unwrap().as_ref(),
+                    &envs,
+                    state.table,
+                    state.wildcard_vars,
+                    false,
+                    state.resolver,
+                );
+                envs.insert(id.clone(), binding_type.clone());
+                linearizer.retype_ident(lin, id, binding_type);
+            }
+
+            stat_map
+                .iter()
+                .try_for_each(|(_, t)| -> Result<(), TypecheckError> {
+                    walk(state, envs.clone(), lin, linearizer.scope(), t)
+                })
```

Comment:
This is not a recursive record so inserting the fields idents into the type environment before type checking the values could cause Issues right? I mean issues like extra completions for the LSP and unbound variables not reported.

Reply:
You're right. I think it wouldn't cause any issue in practice because all records are recursive currently, so probably this match case is just never entered. But still, it's incorrect.

Reply:
Yes and for generated AST it can be an issue

Reply:
But I don't think we'll typecheck a generated AST, though. I mean if you are thinking about the Nix to Nickel compilation.